Make Swift result variables persistent

adrian-prantl opened the pull request:
While also being more efficient and shorter, this is done mainly in preparation for followup changes.

adrian-prantl committed:
Replace std::multimap with llvm::StringMap (NFC)

While also being more efficient and shorter, this is done mainly in
preparation for followup changes.

adrian-prantl commented:
rdar://113352990

adrian-prantl commented:
@swift-ci test

kastiglione reviewed (commented):
Why was this a multimap, and why can it now be a (single) map?

I did a quick read of the places where `equal_range` was used, but on first look it's not clear to me why it was a multimap, nor is it clear what `DeclsAreEquivalent` was used for.

Note: `DeclsAreEquivalent` is no longer used now as of this change.

adrian-prantl commented:
> Why was this a multimap, and why can it now be a (single) map?
> 

I converted a `multimap<String, Decl *>` into a `Stringmap<SmallVector<Decl *>>`

> I did a quick read of the places where `equal_range` was used, but on first look it's not clear to me why it was a multimap, nor is it clear what `DeclsAreEquivalent` was used for.
> 
> Note: `DeclsAreEquivalent` is no longer used now as of this change.

Let me check ...

adrian-prantl commented:
Fix call to the wrong version of SmallVector.erase()

adrian-prantl commented:
@swift-ci test

adrian-prantl committed:
Store CompilerDecls in SwiftPersistenExpressionState::SwiftDeclMap (NFC)

A CompilerDecl is a pair of TypeSystem and Decl *. This is in
preparation for a later patch.

adrian-prantl commented:
Added a second NFC refactoring on top of this patch.

adrian-prantl commented:
@swift-ci test

adrian-prantl commented:
@swift-ci test

adrian-prantl committed:
Move HandloadedModules from SwiftPersistentExprState to SwiftASTContextFE (NFC)

This change is (somewhat surprisingly) NFC, since the list is only
consumed by SwiftExpressionParser and only populated by it as well.
When SwiftASTContextForExpressions is destroyed the persistent state is too.

adrian-prantl committed:
Move the ownership of SwiftPersistentExpressionState to TypeSystemSwiftTypeRefFE

This allows result variables to persist across SwiftASTContexts.
User declarations become invisible in other contexts (NFC)
User imports also don't affect other contexts (NFC)

adrian-prantl commented:
The last patch in the series is still WIP, in the sense of that I still need to add a test for it.

adrian-prantl commented:
@swift-ci test

adrian-prantl commented:
@swift-ci test windows

adrian-prantl commented:
@augusto2112 @kastiglione Would you mind taking another look at this? The goal here is to make Swift result variables portable between different SwiftASTContext instances.

augusto2112 commented on lldb/source/Plugins/ExpressionParser/Swift/SwiftExpressionParser.cpp:
```diff
@@ -401,11 +399,10 @@ class LLDBREPLNameLookup : public LLDBNameLookup {
-                    count, name_const_str.AsCString("<anonymous>"));
-    }
+    if (m_log)
+      m_log->Printf(
```
Nit: you could use `LLDB_LOGF` here

adrian-prantl committed:
Prepopulate the CompilerInvocation with the host plugin search path.

Generally, LLDB takes compiler plugin search paths from serialized
options in Swift modules, however, for the REPL and Playgrounds it is
useful to always add the host toolchain plugins by default.

Unfortunately this is a property of the toolchain lldb is installed
with, so this is not really testable from within an uninstalled build.

rdar://112122752

adrian-prantl commented:
ping :-)

kastiglione commented on lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.h:
```diff
@@ -1037,8 +1014,45 @@ class SwiftASTContextForExpressions : public SwiftASTContext {
+  HandLoadedModuleSet m_hand_loaded_modules;
+
+  /// This returns the list of hand-loaded modules.
+  HandLoadedModuleSet GetHandLoadedModules() { return m_hand_loaded_modules; }
```
Could this be returned as a const reference instead of by copy?

```suggestion
  const HandLoadedModuleSet &GetHandLoadedModules() { return m_hand_loaded_modules; }
```

kastiglione commented on lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.h:
```diff
@@ -1037,8 +1014,45 @@ class SwiftASTContextForExpressions : public SwiftASTContext {
+  typedef llvm::StringMap<swift::AttributedImport<swift::ImportedModule>>
+      HandLoadedModuleSet;
+
+  // This just adds this module to the list of hand-loaded modules, it doesn't
+  // actually load it.
```
It doesn't only _add_ the list of modules, it could overwrites an existing value, if any.

kastiglione commented on lldb/source/Plugins/TypeSystem/Swift/StoringDiagnosticConsumer.h:
```diff
@@ -15,6 +15,8 @@
 
 #include "Plugins/ExpressionParser/Swift/SwiftDiagnostic.h"
 
+#include "lldb/Utility/StreamString.h"
```
minor: seems unrelated to this PR

kastiglione commented on lldb/source/Plugins/ExpressionParser/Swift/SwiftExpressionParser.cpp:
```diff
@@ -248,17 +253,16 @@ class LLDBExprNameLookup : public LLDBNameLookup {
+    if (m_log)
+      m_log->Printf(
+          "[LLDBExprNameLookup::lookupAdditions (%u)] Searching for %s", count,
+          NameStr.empty() ? "<anonymous>" : NameStr.str().c_str());
```
`NameStr` can't be empty at this point, because of the early return guard clause just prior.

kastiglione commented on lldb/source/Plugins/ExpressionParser/Swift/SwiftExpressionParser.cpp:
```diff
@@ -401,34 +409,34 @@ class LLDBREPLNameLookup : public LLDBNameLookup {
+    if (m_log)
+      m_log->Printf(
+          "[LLDBREPLNameLookup::lookupAdditions (%u)] Searching for %s", count,
+          NameStr.empty() ? "<anonymous>" : NameStr.str().c_str());
```
same here, `NameStr` will not be empty.

adrian-prantl committed:
Adopt LLDB_LOG throughout SwiftExpressionParser (NFC)

adrian-prantl commented:
@swift-ci test

adrian-prantl commented:
@swift-ci test windows

adrian-prantl commented:
@swift-ci test

adrian-prantl commented:
@swift-ci test windows

adrian-prantl commented:
@swift-ci test windows
